Implement natural blocks introduced in the Nether Update

Azvyl opened the pull request:
This PR adds several previously unimplemented natural blocks from the Nether Update:
- Crimson Fungus [`minecraft:crimson_fungus`](https://minecraft.wiki/w/Crimson_Fungus)
- Warped Fungus [`minecraft:warped_fungus`](https://minecraft.wiki/w/Warped_Fungus)
- Nether Sprouts [`minecraft:nether_sprouts`](https://minecraft.wiki/w/Nether_Sprouts)
- Crimson Nylium [`minecraft:crimson_nylium`](https://minecraft.wiki/w/Crimson_Nylium)
- Warped Nylium [`minecraft:warped_nylium`](https://minecraft.wiki/w/Warped_Nylium)

This PR does not implement:
- Bone Meal particles when using it on Nylium, Netherrack, and nether fungus

1.16.0 - Nether Update

## Changes
### API changes
- Add new type ids to `BlockTypeIds`
- Add `BlockTypeTags::NYLIUM`, `BlockTypeTags::HUGE_FUNGUS_REPLACEABLE`
- Add `NetherFungus`, `NetherSprouts`, `Nylium`, `NetherTree` classes
- Register the new block to `StringToItemParser`
- Add new `TreeType`: `CRIMSON`, `WARPED`
- Add the new `TreeType` to `TreeFactory`

### Behavioural changes
- NetherRoots and mushrooms get Nylium support
- Using Bone Meal on Netherrack placed next to Nylium will turn the Netherrack into Nylium

## Tests
https://youtu.be/OAmaJGrlAYM

Azvyl committed:
Implement Nether Update nature blocks

Azvyl committed:
uh

Azvyl committed:
Fix planted tree logic

Azvyl committed:
Fix missing Nether Sprouts item

Azvyl committed:
Merge branch 'minor-next' into nether-natural-blocks

Azvyl committed:
Merge branch 'minor-next' into nether-natural-blocks

dktapps commented on src/world/generator/object/NetherTree.php:
```diff
@@ -0,0 +1,196 @@
+
+			for($v = 0; $v < $i; ++$v){
+				$vy = $currentY - $v;
+				if($vy < 0) break;
```
Conditions should always use braces

dktapps commented on src/world/generator/object/TreeFactory.php:
```diff
@@ -37,6 +38,8 @@ public static function get(Random $random, ?TreeType $type = null) : ?Tree{
 			TreeType::ACACIA => new AcaciaTree(),
 			TreeType::BIRCH => new BirchTree($random->nextBoundedInt(39) === 0),
+			TreeType::CRIMSON_FUNGUS_PLANTED => new NetherTree(VanillaBlocks::CRIMSON_STEM(), VanillaBlocks::NETHER_WART_BLOCK(), $random, true, true),
+			TreeType::WARPED_FUNGUS_PLANTED => new NetherTree(VanillaBlocks::WARPED_STEM(), VanillaBlocks::WARPED_WART_BLOCK(), $random, false, true),
```
Please use named parameters for the bools here

dktapps commented on src/world/generator/object/NetherTree.php:
```diff
@@ -0,0 +1,196 @@
+
+	public function canPlaceObject(ChunkManager $world, int $x, int $y, int $z, Random $random) : bool{
+		if($this->planted){
+			return true;
```
I don't like this. Why is the placement logic different when placed by a player vs the generator?

Azvyl replied:
> Huge fungi grow even with blocks above them, and do not replace any solid blocks when grown; however, they do not grow if blocked by the world height limit, or if situated outside of the world.

According to [the wiki](https://minecraft.wiki/w/Huge_fungus#cite_ref-1:~:text=Huge%20fungi%20grow%20even%20with%20blocks%20above%20them%2C%5B1%5D%20and%20do%20not%20replace%20any%20solid%20blocks%20when%20grown%3B%20however%2C%20they%20do%20not%20grow%20if%20blocked%20by%20the%20world%20height%20limit%2C%20or%20if%20situated%20outside%20of%20the%20world.), it turns out it will still grow in both conditions, should I change the function to return `true` or should I just remove the call from `getBlockTransaction()`?

Local test

<img width="2412" height="1080" alt="image" src="https://github.com/user-attachments/assets/56b43b3a-389d-4421-9051-9fda56ed50ca" />

seed: 397832894068853595

dktapps replied:
I'm not sure. It's possible we could rely on `BlockTransaction` validators, though I'm not sure if there's a reason we can't call this code when the player grows the fungus.

dktapps commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,102 @@
+		for($i = 0; $i < 4; ++$i){
+			$pos = $this->position->add($random->nextRange(-1, 1), $random->nextRange(-2, 0), $random->nextRange(-1, 1));
+			$block = $world->getBlock($pos);
+			if($block instanceof Netherrack && !$world->getBlock($pos->up()) instanceof Opaque){
```
`instanceof` is generally frowned upon for blocks

Azvyl replied:
We can compare type id for Netherrack, but for Opaque.. we use `isSolid()`?

dktapps replied:
`!isTransparent()` would be the usual way. Though you may want to question whether that makes sense

dktapps commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,102 @@
+	public function onRandomTick() : void{
+		$world = $this->position->getWorld();
+		$above = $this->getSide(Facing::UP);
+		if($above instanceof Opaque){
```
Same here

dktapps commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,102 @@
+			VanillaBlocks::WARPED_FUNGUS(),
+			VanillaBlocks::WARPED_ROOTS(),
+			VanillaBlocks::NETHER_SPROUTS()
+		];
```
Not sure how I feel about this. I know it's a pattern in older code, but I don't love hardcoding self type ID checks. It makes the class have limited use for plugin custom blocks.

Azvyl replied:
> custom blocks

Are we talking about custom Nylium? this class is `final`, but maybe to work around that we can make this class (and NetherFungus) `abstract`, then add a new class containing Nylium: `getVegetation()`; NetherFungus: `getTreeType()`, `getNyliumTypeId()` for each variant

dktapps replied:
Or parameterising the constructor

Azvyl replied:
> Or parameterising the constructor

That would be better because i wanted to avoid adding new classes

dktapps replied:
Agreed. I prefer only to add classes if new logic is needed. Data should generally be injectable.

dktapps commented on src/block/NetherFungus.php:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	private function grow(?Player $player) : bool{
+		$treeType = $this->getTypeId() === BlockTypeIds::CRIMSON_FUNGUS ? TreeType::CRIMSON_FUNGUS_PLANTED : TreeType::WARPED_FUNGUS_PLANTED;
```
See above comments about hardcoding self type IDs

dktapps commented on src/block/NetherFungus.php:
```diff
@@ -0,0 +1,86 @@
+			$down = $this->getSide(Facing::DOWN);
+			if(
+				($this->getTypeId() === BlockTypeIds::CRIMSON_FUNGUS && $down->getTypeId() === BlockTypeIds::CRIMSON_NYLIUM) ||
+				($this->getTypeId() === BlockTypeIds::WARPED_FUNGUS && $down->getTypeId() === BlockTypeIds::WARPED_NYLIUM)
```
Same here

Azvyl committed:
Apply changes

dktapps commented on src/world/generator/object/NetherTree.php:
```diff
@@ -0,0 +1,198 @@
+
+		$transaction = new BlockTransaction($world);
+		$this->placeStem($x, $y, $z, $random, $this->generateTrunkHeight($random), $transaction);
+		$this->placeHat($x, $y, $z, $random, $transaction);
```
Is there a reason to rename these functions? It looks to me like this logic is identical to the base tree class. Only the stem and hat functions have different names. If we're going to pretend these are trees, we might as well be consistent about it.

Azvyl replied:
I was previously only thinking about naming the helper functions (`placeHatBlock()`, `placeHatDropBlock()`)

Azvyl committed:
Apply changes

dktapps commented on src/block/NetherFungus.php:
```diff
@@ -0,0 +1,95 @@
+			$down = $this->getSide(Facing::DOWN);
+			if($down->getTypeId() === $this->nyliumTypeId){
+				if(mt_rand(1, 100) <= 40){
+					if($this->grow($player)){
```
These conditions could be &&'d together

dktapps commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,100 @@
+		for($i = 0; $i < 4; ++$i){
+			$pos = $this->position->add($random->nextRange(-1, 1), $random->nextRange(-2, 0), $random->nextRange(-1, 1));
+			$block = $world->getBlock($pos);
+			if($block->getTypeId() === BlockTypeIds::NETHERRACK && $world->getBlock($pos->up())->isTransparent()){
```
Beware this will replace stuff like glass, which you probably don't want

Azvyl replied:
It's a check if the netherrack is not covered by a solid opaque block, so I think it will be fine

dktapps replied:
Oh, it's replacing the netherrack itself. My bad

dktapps commented on src/world/generator/object/NetherTree.php:
```diff
@@ -0,0 +1,198 @@
+							}
+						}else{
+							$transaction->addBlockAt($blockX, $blockY, $blockZ, $this->trunkBlock);
+						}
```
This could be simplified to `if(!$isCorner || $random->nextFloat() < 0.1)`, not needing an `else` branch (and could also be merged into the outer condition)

dktapps commented on src/world/generator/object/NetherTree.php:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	protected function canOverride(Block $block) : bool{
+		return $block->canBeReplaced() || $block instanceof NetherFungus || $block instanceof NetherVines;
```
Don't like these `instanceof` checks. Consider using tags instead

Azvyl replied:
`REPLACEABLE_BY_HUGE_FUNGUS`, `HUGE_FUNGUS_REPLACEABLE`, `FUNGUS_REPLACEABLE`, `FUNGUS_GROW_THROUGH`

I'm not sure about the tag naming

Azvyl replied:
<img width="2412" height="1080" alt="image" src="https://github.com/user-attachments/assets/4ece7c86-e766-4576-bba4-e4faa64fe23e" />

<img width="1280" height="720" alt="image" src="https://github.com/user-attachments/assets/92da420a-2de7-4aad-9365-7efc6d2e7be2" />

It turns out that the nether vine is not replaced in bedrock, but is destroyed in java

I didn't implement destroying the block

Azvyl committed:
Apply changes

Azvyl committed:
BlockTypeTags::HUGE_FUNGUS_REPLACEABLE

Azvyl commented on src/world/generator/object/TreeType.php:
```diff
@@ -45,6 +47,8 @@ enum TreeType{
 	case ACACIA;
 	case DARK_OAK;
+	case CRIMSON_FUNGUS_PLANTED;
+	case WARPED_FUNGUS_PLANTED;
```
Do I need to change this?
to `PLANTED_CRIMSON` and `PLANTED_WARPED` perhaps

dktapps commented on src/world/generator/object/NetherTree.php:
```diff
@@ -0,0 +1,191 @@
+		$this->placeCanopy($x, $y, $z, $random, $transaction);
+
+		return $transaction;
+	}
```
I don't think this function is needed anymore if inheritance is used correctly. The logic should be identical to the parent

dktapps commented on src/world/generator/object/NetherTree.php:
```diff
@@ -0,0 +1,191 @@
+					$blockZ = $z + $k;
+
+					if($this->canOverride($transaction->fetchBlockAt($blockX, $blockY, $blockZ))){
+						if(!$isCorner || $random->nextFloat() < 0.1){
```
This can again be merged into the outer condition if you wrap it in (). Also this check is probably faster than the canOverride check, so you could put it before that

dktapps reviewed (commented):
I'm unclear what the intended behaviour is for the override checks. Currently, the fungus will just grow around and through obstructions, which doesn't seem like desired behaviour. More likely it should fail if it encounters a non-overridable block?

dktapps commented:
You could accomplish that by adding a validator to the `BlockTransaction` instead of having lots of `canOverride()` checks

Azvyl commented:
> You could accomplish that by adding a validator to the `BlockTransaction` instead of having lots of `canOverride()` checks

That would make it not grow at all, which is not vanilla behavior. Nether trees will still grow by replacing replaceable blocks (tested with tall grass, fire, glow lichen, resin clump, vine) and replacing nether fungus (regardless of the type of nether tree grown), non-replaceable blocks will remain unchanged

Azvyl committed:
Merge condition

Azvyl committed:
Follow the naming of the rest of the trees

I think I'm still unsure about the naming and purpose of TreeFactory, especially the distinction between player and generator placement.

dktapps commented:
> > You could accomplish that by adding a validator to the `BlockTransaction` instead of having lots of `canOverride()` checks
> 
> That would make it not grow at all, which is not vanilla behavior. Nether trees will still grow by replacing replaceable blocks (tested with tall grass, fire, glow lichen, resin clump, vine) and replacing nether fungus (regardless of the type of nether tree grown), non-replaceable blocks will remain unchanged

Yeah, but then they'll grow through ceilings and get chopped in half by them. That doesn't make sense to me.

dktapps commented on src/world/generator/object/TreeType.php:
```diff
@@ -34,9 +34,9 @@
+ * @method static TreeType PLANTED_WARPED()
  * @method static TreeType SPRUCE()
- * @method static TreeType CRIMSON_FUNGUS_PLANTED()
- * @method static TreeType WARPED_FUNGUS_PLANTED()
```
You don't need to update this doc comment. It's only here for legacy purposes. New usages of new cases should use the native enum syntax.

dktapps commented:
Apparently Bugrock is stupider than I thought, once again:

![Screenshot_20251018_220025_Minecraft](https://github.com/user-attachments/assets/5833cd04-8b19-426c-b253-96989160d15c)

At this point I'm no longer interested in implementing Bugrock behaviour bug for bug and trying to guess whether what they did was intentional or not, on a general level.

This smacks of them forgetting to check for overlapping blocks. I would prefer implementing the Java growth behaviour as described in the wiki so we don't have to constantly guess whether it's correct or not. Eventually Bugrock will no doubt alter the behaviour to match Java after they get dozens of user bug reports over a few years, as with every other parity issue.

The chance of growing a huge fungus is fine to keep, since they can hardly do that by accident...

Azvyl commented:
> I would prefer implementing the Java growth behaviour as described in the wiki so we don't have to constantly guess whether it's correct or not.

Sorry
I don't think I understand what the problem is in the picture, [this](https://minecraft.wiki/w/Huge_fungus#:~:text=Huge%20fungi%20grow%20even%20with%20blocks%20above%20them%2C%5B1%5D%20and%20do%20not%20replace%20any%20solid%20blocks%20when%20grown%3B%20however%2C%20they%20do%20not%20grow%20if%20blocked%20by%20the%20world%20height%20limit%2C%20or%20if%20situated%20outside%20of%20the%20world.) also happens in Java Edition

<img width="1280" height="720" alt="image" src="https://github.com/user-attachments/assets/0dcb343e-b949-4389-b418-98cfa205d7b3" />

> I would prefer implementing the Java growth behaviour as described in the wiki

Which part is missing from this implementation?

dktapps commented:
> > I would prefer implementing the Java growth behaviour as described in the wiki so we don't have to constantly guess whether it's correct or not.
> 
> Sorry I don't think I understand what the problem is in the picture, [this](https://minecraft.wiki/w/Huge_fungus#:~:text=Huge%20fungi%20grow%20even%20with%20blocks%20above%20them%2C%5B1%5D%20and%20do%20not%20replace%20any%20solid%20blocks%20when%20grown%3B%20however%2C%20they%20do%20not%20grow%20if%20blocked%20by%20the%20world%20height%20limit%2C%20or%20if%20situated%20outside%20of%20the%20world.) also happens in Java Edition
> <img alt="image" width="1280" height="720" src="https://private-user-images.githubusercontent.com/67502532/502916086-0dcb343e-b949-4389-b418-98cfa205d7b3.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjA4MzIzNDYsIm5iZiI6MTc2MDgzMjA0NiwicGF0aCI6Ii82NzUwMjUzMi81MDI5MTYwODYtMGRjYjM0M2UtYjk0OS00Mzg5LWI0MTgtOThjZmEyMDVkN2IzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTEwMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUxMDE5VDAwMDA0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg0NmNkYzNmZmZjMzU4ZWU1NjgxNGI3NGYwODE4M2UxYmRiMzYwODVlM2MyOWQzYTdhYmFmZThkYTk4NWU1MWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ulZTnZXYV19glJjuJfPM0_ZQnXxF2DYv6uQRVLOXh8k">
> 
> > I would prefer implementing the Java growth behaviour as described in the wiki
> 
> Which part is missing from this implementation?

Oh, I see. I didn't expect Java to have this level of absurdity too...
In that case it may be done. I'll re-review it tomorrow.

Azvyl committed:
Apply Changes

dktapps commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,100 @@
+					$replace = $world->getBlock($pos);
+					$place = $this->vegetation[$random->nextBoundedInt(count($this->vegetation))];
+					if($replace->getTypeId() === BlockTypeIds::AIR && $place->canBePlacedAt($replace, Vector3::zero(), Facing::DOWN, true)){
+						$tx->addBlock($pos, $place);
```
I'm not sure why this uses a BlockTransaction. It doesn't look like it gets used for anything but setting blocks.

Azvyl replied:
I was thinking about adding a StructureGrowEvent, what do you think?
I'm confused because in Grass there doesn't seem to be an event when using bonemeal.

dktapps replied:
It's an open issue with bone meal in general. For now, it's not used. If we want to add events later, whoever does that can change it to use a transaction.

dktapps commented on src/world/generator/object/NetherTree.php:
```diff
@@ -0,0 +1,177 @@
+		if($random->nextBoundedInt(12) === 0){
+			$i *= 2;
+		}
+		$this->huge = !$this->planted && $random->nextFloat() < 0.06;
```
I think `huge` should be a constructor parameter instead, and have the caller decide whether to use a huge or small one. That would be more plugin-friendly.

dktapps replied:
Pass the height as a constructor parameter too

Azvyl committed:
Apply Changes

Azvyl commented:
Doing a little testing locally, it turns out that the huge variant can still be grown when using bonemeal

![Screenshot_2025-10-20-00-20-53-97_5c8300b655012b1930f2e0a7b81bf6a9](https://github.com/user-attachments/assets/6de4dcbd-0463-4521-81f7-e85e09535107)

dktapps commented:
Yep, seems fine to me

Only thing left seems to be the CS check (it's complaining about an unused import)

Azvyl committed:
Fix CS

ipad54 commented on src/block/NetherFungus.php:
```diff
@@ -0,0 +1,91 @@
+				return true;
+			}
+		}
+		return parent::onInteract($item, $face, $clickVector, $player, $returnedItems);
```
I think this should return false instead of parent call for consistency with other onInteract() overrides

ipad54 commented on src/block/NetherFungus.php:
```diff
@@ -0,0 +1,91 @@
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($item instanceof Fertilizer){
+			$down = $this->getSide(Facing::DOWN);
+			if($down->getTypeId() === $this->nyliumTypeId && mt_rand(1, 100) <= 40 && $this->grow($player)){
```
I don’t think this is correct. In vanilla, bone meal is still subtracted on fail. Also, the chance should not be applied for creative players, vanilla lets them grow the fungus instantly.

Azvyl replied:
I think sapling has this problem too

dktapps replied:
Bone meal is generally not consistent with vanilla at the moment. But that's a topic for another time.

ipad54 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,97 @@
+use pocketmine\utils\Random;
+use function count;
+
+class Nylium extends Block{
```
Shouldn’t this extend Opaque?

ipad54 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,97 @@
+			$this->growVegetation(new Random());
+			return true;
+		}
+		return parent::onInteract($item, $face, $clickVector, $player, $returnedItems);
```
Same about parent call

ipad54 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,97 @@
+					$replace = $world->getBlock($pos);
+					$place = $this->vegetation[$random->nextBoundedInt(count($this->vegetation))];
+					if($replace->getTypeId() === BlockTypeIds::AIR && $place->canBePlacedAt($replace, Vector3::zero(), Facing::DOWN, true)){
+						$world->setBlock($pos, $place);
```
This should check that the position is in the world

ipad54 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,97 @@
+					$pos = $this->position->add($x, 1, $z);
+					$replace = $world->getBlock($pos);
+					$place = $this->vegetation[$random->nextBoundedInt(count($this->vegetation))];
+					if($replace->getTypeId() === BlockTypeIds::AIR && $place->canBePlacedAt($replace, Vector3::zero(), Facing::DOWN, true)){
```
I think canBePlacedAt() check is useless since vegetation can always be supported by nylium

Azvyl replied:
`canBePlacedAt()` is needed because it can grow to blocks other than Nylium

ipad54 replied:
I see

ipad54 commented:
Nylium should be affected by silk touch enchantment

ipad54 commented:
Nether roots, red/brown mushrooms can also be supported by nylium

Azvyl commented:
> This PR does not implement:
> - Nylium support for mushroom and NetherRoots
> - Using Bone Meal on Netherrack placed next to Nylium should convert the Netherrack into Nylium

Do i need to implement that Netherrack part too?

Azvyl committed:
Merge branch 'minor-next' into nether-natural-blocks

ipad54 commented:
> > This PR does not implement:
> > 
> > * Nylium support for mushroom and NetherRoots
> > * Using Bone Meal on Netherrack placed next to Nylium should convert the Netherrack into Nylium
> 
> Do i need to implement that Netherrack part too?

Ah, sorry, I didn’t read the follow up part. You can probably implement all of this in separate PRs if you wish

Azvyl committed:
Apply Changes

Azvyl committed:
phpstan

dktapps commented on src/world/generator/object/TreeType.php:
```diff
@@ -45,6 +45,8 @@ enum TreeType{
+	case CRIMSON;
+	case WARPED;
 	//TODO: cherry blossom, mangrove, azalea
 	//TODO: do crimson and warped "trees" belong here? I'm not sure if they're actually trees or just fungi
```
This TODO can be removed now

Azvyl replied:
To be sure, do i need to delete all three lines or just the last 2 lines?

dktapps replied:
Just this line only

dktapps commented:
Will re-review this on Monday

Azvyl commented:
> > > This PR does not implement:
> > > 
> > > * Nylium support for mushroom and NetherRoots
> > > * Using Bone Meal on Netherrack placed next to Nylium should convert the Netherrack into Nylium
> > 
> > 
> > Do i need to implement that Netherrack part too?
> 
> Ah, sorry, I didn’t read the follow up part. You can probably implement all of this in separate PRs if you wish

I would really like to combine both into this PR, but usually these are made in separate PRs.
What do you think?

dktapps commented:
Those should be easy to do, right? I don't see a reason not to do them here

ShockedPlot7560 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,101 @@
+	/**
+	 * @param Block[] $vegetation An array of Block instances that can be grown on this Nylium block using Bone Meal.
+	 */
+	public function __construct(BlockIdentifier $idInfo, string $name, BlockTypeInfo $typeInfo, private readonly array $vegetation){
```
Unless I am mistaken, I do not see any discussion on this subject. If we keep this, a developer will not be able to add a block to bone meal without overriding the block—which we want to avoid. This may also be the case for grass. Should we modify it here or later to include all other cases?

Azvyl replied:
https://github.com/pmmp/PocketMine-MP/pull/6843#discussion_r2441880618

Sorry I don't quite understand what you mean, `$vegetation` was added to avoid hardcoding self type IDs (`$this->getTypeId() === BlockTypeIds::CRIMSON_NYLIUM`)

ShockedPlot7560 replied:
It avoid to hardcode IDs in the block himself but if a plugin developper want to add their vegetations to nylium, a block override will be needed.

Azvyl replied:
From what i understand, we didn't create it to help modify Nylium, but to add custom (native) blocks

Azvyl replied:
> a block override will be needed.

Are we talking about overriding the block or the class?

dktapps replied:
We do need to rethink how bone meal works generally, but it should come as a more generalised API involving something like a `BoneMealEvent`. But this is outside scope for this PR.

ShockedPlot7560 replied:
Yep, thats not only for this PR, it was a general suggestion not for only this PR scope.

ShockedPlot7560 reviewed (dismissed):
I didn't tested and reviewed the NetherTree structure yet. Will do later if needed

Azvyl committed:
Netherrack & NetherRoots

Azvyl committed:
Remove TreeType comment

DavyCraft648 commented:
~I didn't implement Nylium support to mushrooms because their placement logic was incorrect, they shouldn't be able to be placed directly under the sky.~

Azvyl committed:
Nylium support for mushrooms

Azvyl committed:
Fix bonemeal in survival...

dktapps reviewed (dismissed):
I think this is fine. Barring issues around bone meal handling, but that needs to be more generally reworked anyway and is a topic for another time.

ipad54 commented on src/block/Netherrack.php:
```diff
@@ -23,9 +23,66 @@
+		return false;
+	}
+
+	public function tryTransform() : bool{
```
This should be private

Azvyl committed:
Change visibility

dktapps commented:
FWIW, in the future it'd be better to split up the PR where possible. This PR contains quite a few additions that probably didn't need to be in a single PR. It didn't create problems this time, but might in other PRs.
